src/rgw/rgw_redis_scripts.lua

```diff
@@ -0,0 +1,231 @@
+redis.register_function('assert_lock', assert_lock)
+
+-----
+--- Queue functions
+-----
+
+-- Refer to 'Execution under low memory conditions'
+-- in https://redis.io/docs/latest/develop/interact/programmability/eval-intro/
+---
+
+-- Function to generate a string longer than n by concatenating the charset
+local function generateString(n)
+    local charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890"
+    local result = {}
+    local charset_length = #charset
+
+    local repeats = math.ceil(n / charset_length) + 1
+    for i = 1, repeats do
+        table.insert(result, charset)
+    end
+    return table.concat(result)
+end
+
+--- Add an item to the reserve queue for n bytes
+--- @param keys table A single element list - queue name
+--- @param args table A single-element list - item size
+--- @return number 0 if the item is added to the queue, -lerrorCodes.ENOMEM if the queue is full
+local function reserve(keys, args)
+    local name = "reserve:" .. keys[1]
+    local item_size = tonumber(args[1])
+
+    local randomString = generateString(item_size)
+    --- generate a json of the format {"timestamp": <current time>, "data": <value>}
+    local time = redis.call("TIME")[0]
+    local value = '{"timestamp":' .. redis.call("TIME")[1] .. ',"data":"' .. randomString .. '"}'
+    if not redis.call('LPUSH', name, value) then
+        return formatResponse(-lerrorCodes.ENOMEM, "Not enough memory", "")
+    end
+    return formatResponse(0, "", "")
+end
+
+--- Remove an item from the reserve queue
+--- @param keys table A single element list - queue name
+--- @return number 0 if the item is removed from the queue, -lerrorCodes.ENOENT if the queue is empty
+local function unreserve(keys)
+    local name = "reserve:" .. keys[1]
+
+    local value = redis.call('RPOP', name)
+    if not value then
+        return formatResponse(-lerrorCodes.ENOENT, "Queue is empty", "")
+    end
+    return formatResponse(0, "", "")
+end
+
+--- Commit message to the queue
+--- @param keys table A single element list - queue name
+--- @param args table A single-element list - message
+--- @return number 0 if the message is committed to the queue
+local function commit(keys, args)
+    local name = "queue:" .. keys[1]
+    local message = args[1]
+
+    unreserve(keys)
+    if not redis.call('LPUSH', name, message) then
+        return formatResponse(-lerrorCodes.ENOMEM, "Not enough memory", "")
+    end
+    return formatResponse(0, "", "")
+end
+
+--- Abort the message reservation
+local function abort(keys, args)
+    return unreserve(keys)
+end
+
+--- Read a message from the queue
+--- @param keys table A single element list - queue name
+--- @return string message if the message is read from the queue, nil if the queue is empty
+--- @return This does not remove the message from the queue
+local function read(keys)
+    local name = "queue:" .. keys[1]
+    local value = redis.call('LRANGE', name, -1, -1)[1]
```

Comment:
how many messages you read from the queue?

Reply:
A single message at a time. Should we provide the ability to read messages in batches?

Reply:
yes. i nthe current ckls implementation we can read up to 1000 messages at a time.
this would give much better performance

Reply:
Okay, will allow a `batch_size` arg. Thanks